### ISSUE TITLE
Possibility of hooks for RUN statements ({PRE,POST}RUN)

### ISSUE THREAD

**Issue by LuisAlejandro** [docker/buildx#2087]
### Description

Often, there's the need for executing cleaning or shrinking operations after commands that use or save some sort of cache. This is true specially for package managers (apt, pip, yarn, gems, etc). Today this is solved by manually deleting cache folders, disabling cache or configuring hooks for each package manager. Also, for convenience, these operations are often done on one-liners, making readability more difficult and reducing the flexibility of the Dockerfile.

My proposal is to implement `PRERUN` and `POSTRUN` hooks that would execute for every `RUN` statement. The commands specified in the hooks would be executed in the same layer of the original statement. In practice it would be like prepending and appending commands to the existent one. I think this would simplify many Dockerfiles that implement one-liners to make the shrinking of layers effective.

What do you think?

**Comment by jsternberg** [comment]
I'm not completely sure I understand the proposal. Would `PRERUN` and `POSTRUN` run only within the layer that they are next to or for every `RUN` statement in the dockerfile?

I suspect this solution might make things either harder to understand or cause more problems than it would help.

If the `PRERUN` and `POSTRUN` are after every command, we'd have to make every `RUN` command include every statement in a `PRERUN` and `POSTRUN`. So something like this:

```
POSTRUN rm -rf /var/lib/apt/lists/*
RUN apt-get update && apt-get install -y golang
WORKDIR /src
RUN --mount=/src go build
```

The `rm` to remove the apt-get cache would happen for both run commands which would cause the second run to be `go build && rm -rf /var/lib/apt/lists/*` which I think would be undesirable.

If this was only associated with the original `RUN` command, I think it would be a bit hard to follow personally.

I do think I understand the underlying frustration. When I write a dockerfile, it can feel like I have to go find another dockerfile with an `apt-get` command so I can find the correct cleanup command to follow "best practices". Unfortunately, the dockerfile frontend may not be a good place to try and address this. The dockerfile frontend is intentionally a rather "raw" frontend where the commands being run are fairly explicit and non-magical.

Buildkit itself has a way to customize the frontend. One proof of concept for this is [typebuild](https://typebuild.rumpl.dev/) which exposes the buildkit LLB format to typescript and then you can write typescript as a high-level language. For something like `apt`, you could have something like `new Apt(packages).install()` and it could capture the pre-run and post-run of the apt command specifically.

**Comment by jsternberg** [comment]
I'm going to close this issue for now. I don't think we presently have any plan to implement this. If you'd like to continue the discussion, I can reopen it or we can open another issue to start discussing other ways to make dockerfiles easier to create in these circumstances.
